dao/src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,101 @@
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Component;
+
+import java.sql.PreparedStatement;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+@Component
+public class CertificateDaoImpl implements CertificateDao {
+
+    private static final String UPDATE_CERTIFICATE = "UPDATE gift_certificate  SET " +
+            "name = COALESCE(?, name), description = COALESCE(?, description), " +
+            "price = COALESCE(?, price), duration = COALESCE(?, duration) WHERE (`id` = ?)";
+    private static final String INSERT_INTO_CERT_HAS_TAG = "INSERT INTO gift_certificate_has_tag " +
+            "(`gift_certificate_id`, `tag_id`) VALUES (?, ?)";
+    private static final String SELECT_BY_SEARCH_PARAMETERS = "SELECT c.id, c.name, c.description, c.price, " +
+            "c.duration, c.create_date, c.last_update_date " +
+            "FROM gift_certificate AS c " +
+            "JOIN gift_certificate_has_tag AS cht ON c.id = cht.gift_certificate_id " +
+            "JOIN tag AS t on cht.tag_id=t.id ";
+    private static final String INSERT_CERTIFICATE = "INSERT INTO gift_certificate " +
+            "(`name`, `description`, `price`, `duration`) VALUES (?, ?, ?, ?)";
+    private static final String SELECT_CERTIFICATE_BY_ID = "SELECT * FROM gift_certificate WHERE id=?";
+    private static final String SELECT_ALL_CERTIFICATES = "SELECT * FROM gift_certificate";
+    private static final String DELETE_CERTIFICATE = "DELETE FROM gift_certificate WHERE id = ?";
+    private static final String COLUMN_ID = "id";
+
+    private final JdbcTemplate jdbcTemplate;
+    private final CertificateMapper certificateMapper;
+
+    public CertificateDaoImpl(JdbcTemplate jdbcTemplate, CertificateMapper certificateMapper) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.certificateMapper = certificateMapper;
+    }
+
+    @Override
+    public Integer updateCertificate(Certificate certificate) {
+        return jdbcTemplate.update(UPDATE_CERTIFICATE,
+                certificate.getName(),
+                certificate.getDescription(),
+                certificate.getPrice(),
+                certificate.getDuration(),
+                certificate.getId());
+    }
+
+    @Override
+    public Integer createCertificateHasTag(Long certificateId, Long tagId) {
+        return jdbcTemplate.update(INSERT_INTO_CERT_HAS_TAG, certificateId, tagId);
+    }
+
+    @Override
+    public List<Certificate> searchByParameters(String query, List<String> parameters) {
+        String searchQuery = SELECT_BY_SEARCH_PARAMETERS.concat(query);
+        return jdbcTemplate.query(searchQuery, certificateMapper, parameters.toArray());
+    }
+
+    @Override
+    public Long createFromEntity(Certificate entity) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        try {
+
+            jdbcTemplate.update(con -> {

```

Comment:
It's not recommended to use abbreviations and short name anywhere. Even if they looks obviously.

Reply:
thanks. i'll remember. fixed